### PR TITLE
pyramid: Fix which package is the correct caller in _traced_init.

### PR DESCRIPTION
**Pull request by kenrobbins**
# Description

The _traced_init wrapper is getting one caller too far up. Example below.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

I put a pdb statement in the Registry constructor, which is called by the Configurator constructor. This is the same code running. The top has otel enabled and the bottom has it disabled.

```
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/paste/deploy/loadwsgi.py(152)invoke()
-> return fix_call(context.object, context.global_conf, **context.local_conf)
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/paste/deploy/util.py(55)fix_call()
-> val = callable(*args, **kw)
  /opt/src/webapp/webapp/__init__.py(44)main()
-> config = Configurator(root_factory=Root, settings=settings)
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/opentelemetry/instrumentation/pyramid/__init__.py(142)_traced_init()
-> wrapped(*args, **kwargs)
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/pyramid/config/__init__.py(314)__init__()
-> registry = Registry(self.package_name)
> /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/pyramid/registry.py(62)__init__()
-> Components.__init__(self, package_name, *args, **kw)
(Pdb) package_name
'paste.deploy'

 /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/paste/deploy/loadwsgi.py(152)invoke()
-> return fix_call(context.object, context.global_conf, **context.local_conf)
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/paste/deploy/util.py(55)fix_call()
-> val = callable(*args, **kw)
  /opt/src/webapp/webapp/__init__.py(44)main()
-> config = Configurator(root_factory=Root, settings=settings)
  /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/pyramid/config/__init__.py(314)__init__()
-> registry = Registry(self.package_name)
> /home/ken/.pyenv/versions/3.7.5/envs/webapp/lib/python3.7/site-packages/pyramid/registry.py(62)__init__()
-> Components.__init__(self, package_name, *args, **kw)
(Pdb) package_name
'webapp'
``` 

Added unit test to ensure registry name value matches the module of the caller.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [X] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [X] Followed the style guidelines of this project
- [X] Changelogs have been updated
- [X] Unit tests have been added
- [X] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py [hunk 1]
```diff
@@ -134,10 +134,12 @@ def _traced_init(wrapped, instance, args, kwargs):
...
-        # Get the package for the third frame up from this one.
-        # Default is `level=2` which will give us the package from `wrapt`
-        # instead of the desired package (the caller)
-        kwargs["package"] = caller_package(level=3)
+        # Get the package for the 2nd frame up from this one.
+        # Default is `level=2` one level down (in Configurator.__init__).
+        # We want the 3rd level from _there_. Since we are already 1 level above,
+        # we need the 2nd level up from here, which will give us the package from
+        # `wrapt` instead of the desired package (the caller)
+        kwargs["package"] = caller_package(level=2)
```
Comment (codeboten) [inline instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py:142]: Could a test be added to validate this?

**Review by codeboten — COMMENTED** [review]
Thanks for the contribution @kenrobbins! Please sign the CLA

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,12 +9,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- `opentelemetry-exporter-pyramid` Fixed which package is the correct caller in _traced_init.
```
Comment (lzchen) [inline CHANGELOG.md:12]:
```suggestion
- `opentelemetry-instrumentation-pyramid` Fixed which package is the correct caller in _traced_init.
```

File: instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py (hunk 1 above)
Comment (kenrobbins) [reply instrumentation/opentelemetry-instrumentation-pyramid/src/opentelemetry/instrumentation/pyramid/__init__.py:142]: done

**Comment by kenrobbins** [comment]
CLA is signed. Tests added. Should be mergeable now I think.

**Review by lzchen — APPROVED** [review]

**Comment by ocelotl** [comment]
@kenrobbins please fix lint :v:

**Comment by srikanthccv** [comment]
@kenrobbins Please fix the lint and conflict

**Comment by kenrobbins** [comment]
@ocelotl @srikanthccv please check now. should be fixed. thanks.

**Comment by srikanthccv** [comment]
@kenrobbins requires some re-formatting to fix lint and we are good to go.

**Comment by kenrobbins** [comment]
@ocelotl @srikanthccv fixed lint again. 🤞  thanks.

**Comment by lzchen** [comment]
@kenrobbins 
Looks like there is a test that is failing?

**Comment by kevellis124** [comment]
So I've been trying to see what's going on with the new test that is failing.
@lzchen it looks like the actual `_trace_init` func for pyramid was perhaps designed to be run on 
Looks like the assumption in the code is that the `caller_package` needs to look a certain amount of frames up to get the caller. It defaults to `2` which works for CPython implementations of Python, but when using Pypy, there is another `Wrapt` frame which requires looking above another frame.

Current working theory:
Current code is broken for CPython but works for Pypy

Kens fix fixes it for CPython but breaks it for Pypy

Possible fix: conditionally check for which interpreter is being used in `_traced_init` to decide how many arbitrary frames we need to look through

**Comment by kevellis124** [comment]
```
import sys
if "PyPy" in sys.version:
  level=3
else:
  level=2
```

**Comment by kenrobbins** [comment]
I committed a change to account for that. Let's see if it passes.

**Comment by kevellis124** [comment]
@srikanthccv if you have a second to approve the workflows again we'd appreciate it 🙏

**Comment by kevellis124** [comment]
Once more please @srikanthccv 😬 thanks

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**